### PR TITLE
feature(transpiler): add key to transpile method

### PR DESCRIPTION
**Pull request by jacksonopp**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngneat/transloco/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
  - I may need help with testing since this feature doesn't add any functionality directly, but it does not break any existing tests afaik
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

https://github.com/ngneat/transloco/issues/568

The `TranslocoTranspiler.transpile` does not give the key of the current translation.

Issue Number: 568

## What is the new behavior?

The `TranslocoTranspiler.transpile` now accepts the key as a parameter, and the `TranslocoService.translate` method now passes in the key.

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

## Other information
This is one of my first contributions to an open source project, so I expect that I will need to make some changes.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -123,4 +123,3 @@
     }
   }
 }
-
```
Comment (shaharkazaz) [inline package.json:126]: Please revert the changes in the package and package lock.

File: docs/yarn.lock [hunk 2]
```diff
@@ -2681,7 +2681,7 @@ class-utils@^0.3.5:
     isobject "^3.0.0"
```
Comment (shaharkazaz) [inline docs/yarn.lock:2681]: Please revert.

File: docs/docs/transpiler.mdx [hunk 3]
```diff
@@ -112,7 +112,7 @@ You can also provide a custom transpiler by creating a class that implements the
 import { TranslocoTranspiler } from '@ngneat/transloco';
 
 export class CustomTranspiler implements TranslocoTranspiler {
-  transpile(value: any, params, translation: Translation) {
+  transpile(value: any, params, translation: Translation, key?: any) {
```
Comment (shaharkazaz) [inline docs/docs/transpiler.mdx:115]: Why is this optional and typed as any?

File: apps/transloco-playground/src/assets/i18n/custom-transpilers/es.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (shaharkazaz) [inline apps/transloco-playground/src/assets/i18n/custom-transpilers/es.json:1]: I don't see the point in adding this to the docs, this is a very specific implementation you needed.

File: libs/transloco/src/lib/transloco.transpiler.ts [hunk 5]
```diff
@@ -16,7 +16,7 @@ import {
 export const TRANSLOCO_TRANSPILER = new InjectionToken('TRANSLOCO_TRANSPILER');
 
 export interface TranslocoTranspiler {
-  transpile(value: any, params: HashMap, translation: Translation): any;
+  transpile(value: any, params: HashMap, translation: Translation, key?: any): any;
```
Comment (shaharkazaz) [inline libs/transloco/src/lib/transloco.transpiler.ts:19]: Why is this optional and typed as any? Please add the following comment above:
```
// TODO Change parameters to object in the next major release
```

File: package.json (hunk 1 above)
Comment (jacksonopp) [reply package.json]: This has been reverted

File: docs/yarn.lock (hunk 2 above)
Comment (jacksonopp) [reply docs/yarn.lock]: This has been reverted

File: libs/transloco/src/lib/transloco.transpiler.ts (hunk 5 above)
Comment (jacksonopp) [reply libs/transloco/src/lib/transloco.transpiler.ts]: Added TODO.

I added `key` as optional since the built in transpilers don't consume it, and it allowed me to change less of the core library. However after thinking about it more I understand better how the transpilers are hooked up, so I have removed the optional type, and am now passing in 'key' wherever `transpile` is being called.

This does mean that I had to update the specs to now pass in a key. I have not added any specs since because this isn't really adding any functionality to the built in transpilers, i'm not sure what exactly I should be testing. I will definitely need help understanding what I should be looking for in the specs.

As for the `any` type, I was following the type of the `value` parameter. Looking at the `translocoService.translate` method, `key` should be `TranslateParams` correct?

File: docs/docs/transpiler.mdx (hunk 3 above)
Comment (jacksonopp) [reply docs/docs/transpiler.mdx]: Updated to non-optional. 

More discussion here: https://github.com/ngneat/transloco/pull/569#discussion_r887043400

File: libs/transloco/src/lib/transloco.transpiler.ts (hunk 5 above)
Comment (shaharkazaz) [reply libs/transloco/src/lib/transloco.transpiler.ts]: @jacksonopp I don't think we need to add any specs here, it's just passing another param.
The key type is `string`

File: apps/transloco-playground/src/assets/i18n/custom-transpilers/es.json (hunk 4 above)
Comment (shaharkazaz) [reply apps/transloco-playground/src/assets/i18n/custom-transpilers/es.json:1]: I think you can make a simpler example by providing a custom transpiler that logs the paramaters.

**Comment by shaharkazaz** [comment]
@jacksonopp fix the types please and force push all the changes in one commit made using `yarn commit`/`npm rnu commit`.

File: docs/docs/plugins/message-format.mdx [hunk 6]
```diff
@@ -18,7 +18,7 @@ npm i messageformat @ngneat/transloco-messageformat
 ## Usage
 The `MessageFormatTranspiler` is compatible with the `DefaultTranspiler` and therefore you can switch without worry that it will break your current translations.
 
-It then enables support for the following within youIt enables support for the following within your i18n translation files:
+It then enables support for the following within your i18n translation files:
```
Comment (shaharkazaz) [inline docs/docs/plugins/message-format.mdx:21]: Are you updated with master? this was already fixed

File: apps/transloco-playground/src/app/custom-transpilers/custom-transpilers.component.html [hunk 7]
```diff
@@ -0,0 +1,17 @@
+<h2>Custom Transpilers</h2>
+<p>In order to test this, you will need to provide the <code>customTranspiler</code> in app.module.ts</p>
```
Comment (shaharkazaz) [inline apps/transloco-playground/src/app/custom-transpilers/custom-transpilers.component.html:2]: Provide the custom transpiler in the app module, it's just logging the params now 🙂 
I don't think we need a dedicated page.

File: apps/transloco-playground/src/app/custom-transpilers/custom.transpiler.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import { HashMap, Translation, TranslocoTranspiler, TRANSLOCO_TRANSPILER } from "@ngneat/transloco";
+
+export class CustomTranspiler implements TranslocoTranspiler {
+  transpile(value: any, params: HashMap<any>, translation: Translation, key: any) {
+    console.log({value, params, translation, key})
+    return value
+  }
+}
```
Comment (shaharkazaz) [inline apps/transloco-playground/src/app/custom-transpilers/custom.transpiler.ts:8]:
```suggestion
export class CustomTranspiler implements TranslocoTranspiler {
  transpile(value: any, params: HashMap<any>, translation: Translation, key: any) {
    console.log('Custom Transpiler called with:', {value, params, translation, key});
    
    return value;
  }
}
```

File: docs/yarn.lock (hunk 2 above)
Comment (shaharkazaz) [reply docs/yarn.lock:2681]: Seems like it's still here

File: libs/transloco/src/lib/transloco.transpiler.ts [hunk 9]
```diff
@@ -81,7 +82,8 @@ export class DefaultTranspiler implements TranslocoTranspiler {
   protected handleObject(
     value: any,
     params: HashMap = {},
-    translation: Translation
+    translation: Translation,
+    key: any
```
Comment (shaharkazaz) [inline libs/transloco/src/lib/transloco.transpiler.ts:86]: why any?

File: libs/transloco/src/lib/transloco.transpiler.ts [hunk 10]
```diff
@@ -104,9 +106,10 @@ export class DefaultTranspiler implements TranslocoTranspiler {
   protected handleArray(
     value: string[],
     params: HashMap = {},
-    translation: Translation
+    translation: Translation,
+    key: any
```
Comment (shaharkazaz) [inline libs/transloco/src/lib/transloco.transpiler.ts:110]: why any?

File: libs/transloco/src/lib/transloco.transpiler.ts [hunk 11]
```diff
@@ -147,7 +150,7 @@ export class FunctionalTranspiler
     super();
   }
 
-  transpile(value: any, params: HashMap = {}, translation: Translation): any {
+  transpile(value: any, params: HashMap = {}, translation: Translation, key: any): any {
```
Comment (shaharkazaz) [inline libs/transloco/src/lib/transloco.transpiler.ts:153]: why any?

File: docs/yarn.lock (hunk 2 above)
Comment (jacksonopp) [reply docs/yarn.lock:2681]: 🤦 I'll fix it again

Comment (jacksonopp) [reply docs/yarn.lock]: This should be fixed again. 

Thank you for your patience as I work thru my first OSS contribution by the way!

File: docs/docs/plugins/message-format.mdx (hunk 6 above)
Comment (jacksonopp) [reply docs/docs/plugins/message-format.mdx]: This should be resolved now

File: apps/transloco-playground/src/app/custom-transpilers/custom-transpilers.component.html (hunk 7 above)
Comment (jacksonopp) [reply apps/transloco-playground/src/app/custom-transpilers/custom-transpilers.component.html]: Removed page

File: apps/transloco-playground/src/app/custom-transpilers/custom.transpiler.ts (hunk 8 above)
Comment (jacksonopp) [reply apps/transloco-playground/src/app/custom-transpilers/custom.transpiler.ts]: Good suggestion, I have made this change

File: libs/transloco/src/lib/transloco.transpiler.ts (hunk 9 above)
Comment (jacksonopp) [reply libs/transloco/src/lib/transloco.transpiler.ts]: Changed to `string`

File: libs/transloco/src/lib/transloco.transpiler.ts (hunk 10 above)
Comment (jacksonopp) [reply libs/transloco/src/lib/transloco.transpiler.ts]: Changed to `string`

File: libs/transloco/src/lib/transloco.transpiler.ts (hunk 11 above)
Comment (jacksonopp) [reply libs/transloco/src/lib/transloco.transpiler.ts]: Changed to `string`

**Comment by jacksonopp** [comment]
Thank you for your patience as I work thru my first OSS contribution by the way!

File: apps/transloco-playground/src/app/custom.transpiler.ts [hunk 12]
```diff
@@ -0,0 +1,13 @@
+import { HashMap, Translation, TranslocoTranspiler, TRANSLOCO_TRANSPILER } from "@ngneat/transloco";
+
+export class CustomTranspiler implements TranslocoTranspiler {
+  transpile(value: any, params: HashMap<any>, translation: Translation, key: any) {
```
Comment (shaharkazaz) [inline apps/transloco-playground/src/app/custom.transpiler.ts:4]:
```suggestion
  transpile(value: any, params: HashMap<any>, translation: Translation, key: string) {
```

Comment (jacksonopp) [reply apps/transloco-playground/src/app/custom.transpiler.ts]: Fixed

**Comment by shaharkazaz** [comment]
@jacksonopp please note that the CI is failing

**Comment by jacksonopp** [comment]
@shaharkazaz 

### Looking into the unit tests
Looks like we're seeing a peer dependency mismatch around rxjs. I think running an npm install should fix this, and would involve committing a change to package-lock.json. Is that okay with you?

### Looking into the e2e tests, I'm seeing the following failures:

- `locale.spec.ts` is failing because it is displaying the wrong unix date, and it fails regardless of whether or not we are using the custom transpiler. I was playing around with changing my time zone, and that seemed to affect the displayed output. Adding `timeZone: 'UTC'` as a parameter to the `translocoDate` pipe normalizes the timezone, and seems to solve this issue. I'm not sure what has changed to cause this. Does this sound correct to you?

- `full-cycle.spec.ts` is failing because the custom transpiler just returns the value without doing any transpiling. Removing the custom transpiler from the providers array seems to fix this issue. Since we've updated the documentation site with the new `key` input, I think its fine if we do not include this in the transloco playground. It wasn't there before anyway.

Thoughts?

**Comment by shaharkazaz** [comment]
@jacksonopp 
* You didn't make any change on the package so I'm not sure why would that change anything 🤔 But let's try it, make sure you have node 16+ and npm 8+.
* Let's try to remove the custom compiler from the playground.
* I don't think there was any change in the locale area, and last couple of commits passed, I'll take a look when I get time if the issue persists, anyways you didn't cause the issue 🙂

**Comment by jacksonopp** [comment]
@shaharkazaz 

- Updated package-lock.json. I used `node v16.15.1` and `npm 8.11.0`
- Removed the custom transpiler from the playground
- Running through the specs, I'm now seeing the unit tests around locales failing for the same reason as the e2e tests, but sounds like you're going to investigate it.

**Comment by jacksonopp** [comment]
@shaharkazaz Any update on this? Looks like all specs are passing now.

**Comment by shaharkazaz** [comment]
@jacksonopp sorry man I'm swamped, I'll get to it as soon as I have some free time to do one last review 🙏

**Comment by jacksonopp** [comment]
@shaharkazaz I understand you are probably busy, but any more updates?

**Comment by shaharkazaz** [comment]
@jacksonopp Sorry man for disappearing, had a lot of stuff going on and just couldn't find the time.
If you'd like we can pick this up again.

**Comment by jacksonopp** [comment]
@shaharkazaz All good. I've updated my fork to match master here.

**Comment by shaharkazaz** [comment]
@jacksonopp nice work on this man. I hope to see you contributing more 😄
